pkg/ingestor/parser/csaf/parser_csaf.go

```diff
@@ -0,0 +1,252 @@
+		purl := findPurl(ctx, b, product_ref)
+		if purl != nil {
+			return purl
+		}
+	}
+
+	return nil
+}
+
+func findProductRef(ctx context.Context, tree csaf.ProductBranch, product_id string) *string {
+	for _, r := range tree.Relationships {
+		if r.FullProductName.ID == product_id {
+			return &r.ProductRef
+		}
+	}
+
+	for _, b := range tree.Branches {
+		pref := findProductRef(ctx, b, product_id)
+		if pref != nil {
+			return pref
+		}
+	}
+	return nil
+}
+
+func findActionStatement(tree *csaf.Vulnerability, product_id string) *string {
+	for _, r := range tree.Remediations {
+		for _, p := range r.ProductIDs {
+			if p == product_id {
+				return &r.Details
+			}
+		}
+	}
+	return nil
+}
+
+func findImpactStatement(tree *csaf.Vulnerability, product_id string) *string {
+	for _, t := range tree.Threats {
+		if t.Category == "impact" {
+			for _, p := range t.ProductIDs {
+				if p == product_id {
+					return &t.Details
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (c *csafParser) findPkgSpec(ctx context.Context, product_id string) (*generated.PkgInputSpec, error) {
+	pref := findProductRef(ctx, c.csaf.ProductTree, product_id)
+	if pref == nil {
+		return nil, fmt.Errorf("unable to locate product reference for id %s", product_id)
+	}
+
+	purl := findPurl(ctx, c.csaf.ProductTree, *pref)
+	if purl == nil {
+		return nil, fmt.Errorf("unable to locate product url for reference %s", *pref)
+	}
+
+	return helpers.PurlToPkg(*purl)
+}
+
+func (c *csafParser) generateVexIngest(ctx context.Context, cve *generated.CVEInputSpec, vuln *csaf.Vulnerability, status string, product_id string) *assembler.VexIngest {
+	logger := logging.FromContext(ctx)
+	vi := &assembler.VexIngest{}
+
+	vd := generated.VexStatementInputSpec{}
+	vd.KnownSince = c.csaf.Document.Tracking.CurrentReleaseDate
+	vd.Origin = c.csaf.Document.Tracking.ID
+	vd.VexJustification = generated.VexJustificationNotProvided
+
+	if vexStatus, ok := vexStatusMap[status]; ok {
+		vd.Status = vexStatus
+	}
+
+	var statement *string
+	if vd.Status == generated.VexStatusNotAffected {
+		statement = findImpactStatement(vuln, product_id)
+	} else {
+		statement = findActionStatement(vuln, product_id)
+	}
+
+	if statement != nil {
+		vd.Statement = *statement
+	}
+
+	for _, flag := range vuln.Flags {
+		found := false
+		for _, pid := range flag.ProductIDs {
+			if pid == product_id {
+				found = true
+			}
+		}
+		if found {
+			if just, ok := justificationsMap[flag.Label]; ok {
+				vd.VexJustification = just
+			}
+		}
+	}
+
+	vi.VexData = &vd
+	vi.CVE = cve
+
+	pkg, err := c.findPkgSpec(ctx, product_id)
+	if err != nil {
+		logger.Warnf("[csaf] unable to locate package for not-affected product %s", product_id)
+		return nil
+	}
+	vi.Pkg = pkg
+
+	return vi
+}
+
+func (c *csafParser) GetPredicates(ctx context.Context) *assembler.IngestPredicates {
+	rv := &assembler.IngestPredicates{}
+	var vis []assembler.VexIngest
+	var cvs []assembler.CertifyVulnIngest
+
+	if len(c.csaf.Vulnerabilities) > 0 {
```

Comment:
nit: can remove check here since range handles it, unless the intent is to have a special case to add `CertifyVuln` to `NoVuln` in the special case